src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/ElasticSearchDAOQuerySupport.scala

```diff
@@ -0,0 +1,124 @@
+package org.broadinstitute.dsde.firecloud.dataaccess
+
+import org.broadinstitute.dsde.firecloud.model._
+import org.broadinstitute.dsde.firecloud.model.ModelJsonProtocol._
+import org.elasticsearch.search.aggregations.{AggregationBuilders, Aggregations}
+import org.elasticsearch.client.transport.TransportClient
+import org.elasticsearch.action.search.{SearchRequest, SearchRequestBuilder, SearchResponse}
+import org.elasticsearch.search.aggregations.bucket.terms.Terms
+import spray.json._
+import spray.json.DefaultJsonProtocol._
+import scala.collection.JavaConverters._
+import scala.concurrent.Future
+import scala.concurrent.ExecutionContext.Implicits.global
+
+
+trait ElasticSearchDAOQuerySupport extends ElasticSearchDAOSupport {
+
+  def createListOfMusts(fields: Map[String, Seq[String]]): Seq[QueryMap] = {
+    (fields map {
+      case (k, v) => ESBool(createESShouldForTerms(k, v))
+    }).toSeq
+  }
+
+  def createESShouldForTerms(attribute: String, terms: Seq[String]): ESShould = {
+    val clauses: Seq[ESTerm] = terms map {
+      case (term: String) => ESTerm(Map(attribute -> term))
+    }
+    ESShould(clauses)
+  }
+
+  def createQuery(criteria: LibrarySearchParams): Seq[QueryMap] = {
+    (criteria.searchString, criteria.searchFields.size) match {
+      case (None | Some(""), 0) => Seq(new ESMatchAll)
+      case (None | Some(""), _) => createListOfMusts(criteria.searchFields)
+      case (Some(searchTerm: String), 0) => Seq(new ESMatch(searchTerm.toLowerCase))
+      case (Some(searchTerm: String), _) => createListOfMusts(criteria.searchFields) :+ new ESMatch(searchTerm.toLowerCase)
+    }
+  }
+
+  def addAggregationsToQuery(searchReq: SearchRequestBuilder, aggFields: Seq[String], maxAggs: Option[Int]): SearchRequestBuilder = {
+    aggFields foreach { field: String =>
+      val terms = AggregationBuilders.terms(field)
+      if (maxAggs.isDefined) {
+        terms.size(maxAggs.get)
+      }
+      searchReq.addAggregation(terms.field(field + ".raw"))
+    }
+    searchReq
+  }
+
+  def createESSearchRequest(client: TransportClient, indexname: String, qmseq: Seq[QueryMap], from: Int, size: Int): SearchRequestBuilder = {
+    client.prepareSearch(indexname)
+      .setQuery(ESConstantScore(ESFilter(ESBool(ESMust(qmseq)))).toJson.compactPrint)
```

Comment:
I'm wondering if we should be generating the json ourselves, or if we should be using the client, a la https://www.elastic.co/guide/en/elasticsearch/client/java-api/2.4/java-term-level-queries.html#java-query-dsl-term-query. Let's chat in person about pros/cons?